README.md
# auth-request

auth-request allows you to add access control to your HTTP services based
on a subrequest to a configured haproxy backend. The workings of this Lua
script are loosely based on the [ngx_http_auth_request_module][1] module
for nginx.

## Set-Up

1. Load this Lua script in the `global` section of your `haproxy.cfg`:
```
global
	# *snip*
	lua-load /usr/share/haproxy/auth-request.lua
```
2. Define a backend that is used for the subrequests:
```
backend auth_request
	mode http
	server auth_request 127.0.0.1:8080 check
```
3. Execute the subrequest in your frontend (as early as possible):
```
frontend http
	mode http
	listen :::80 v4v6
	# *snip*

	#                             Backend name     Path to request
	http-request lua.auth-request auth_request     /is-allowed
```
4. Act on the results:
```
frontend http
	# *snip*
	
	http-request deny if ! { var(txn.auth_response_successful) -m bool }
```

## The Details

The Lua script will make a HTTP request to the *first* server in the given
backend that is *not* marked as `DOWN`. This allows for basic health checking
of the auth-request backend. If you need more complex processing of the
request forward the auth-request to a separate haproxy *frontend* that
performs the required modifications to the request and response.

The requested URL is the one given in the second parameter.

Any request headers will be forwarded as-is to the auth-request backend.

The Lua script will define the `txn.auth_response_successful` variable as
true iff the subrequest returns an HTTP status code of `2xx`. The status code
of the subrequest will be returned in `txn.auth_response_code`. If the
subrequest does not return a valid HTTP response the status code is set
to `500 Internal Server Error`.

## Known limitations

- The Lua script only supports basic health checking, without redispatching
  or load balancing of any kind.
- The response headers of the subrequest are not exposed outside the script.
- The backend must not be using TLS.
- IPv6 is only supported in haproxy 1.8.4+ (not released as of 2018-01-15), due
  [do a bug][2]. You can [monkeypatch by reverting commit 57950b4][3].

[1]: http://nginx.org/en/docs/http/ngx_http_auth_request_module.html
[2]: http://git.haproxy.org/?p=haproxy-1.8.git;a=commit;h=9db449a701cd9e43a04f49e2e477193fa5636323
[3]: https://github.com/TimWolla/haproxy-auth-request/commit/57950b4639542ba429e54b959604e33237c6cffe


debian/changelog
haproxy-auth-request (0) UNRELEASED; urgency=medium

  * Initial Release.

 -- Tim Düsterhus <tim@bastelstu.be>  Mon, 08 Jan 2018 22:56:57 +0000


auth-request.lua
-- The MIT License (MIT)
--
-- Copyright (c) 2018 Tim Düsterhus
--
-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the "Software"), to deal
-- in the Software without restriction, including without limitation the rights
-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-- copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:
--
-- The above copyright notice and this permission notice shall be included in all
-- copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.

local http = require("socket.http")

--- Monkey Patches around bugs in haproxy's Socket class
-- This function calls core.tcp(), fixes a few methods and
-- returns the resulting socket.
-- @return Socket
function create_sock()
	local sock = core.tcp()

	-- https://www.mail-archive.com/haproxy@formilux.org/msg28574.html
	sock.old_receive = sock.receive
	sock.receive = function(socket, pattern, prefix)
		local a, b
		if pattern == nil then pattern = "*l" end
		if prefix == nil then
			a, b = sock:old_receive(pattern)
		else
			a, b = sock:old_receive(pattern, prefix)
		end
		return a, b
	end

	-- https://www.mail-archive.com/haproxy@formilux.org/msg28604.html
	sock.old_settimeout = sock.settimeout
	sock.settimeout = function(socket, timeout)
		socket:old_settimeout(timeout)

		return 1
	end

	return sock
end

core.register_action("auth-request", { "http-req" }, function(txn, be, path)
	txn:set_var("txn.auth_response_successful", false)

	-- Check whether the given backend exists.
	if core.backends[be] == nil then
		txn:Alert("Unknown auth-request backend '" .. be .. "'")
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- Check whether the given backend has servers that
	-- are not `DOWN`.
	local addr = nil
	for name, server in pairs(core.backends[be].servers) do
		if server:get_stats()['status'] ~= "DOWN" then
			addr = server:get_addr()
		end
	end
	if addr == nil then
		txn:Warning("No servers available for auth-request backend: '" .. be .. "'")
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- Transform table of request headers from haproxy's to
	-- socket.http's format.
	local headers = {}
	for header, values in pairs(txn.http:req_get_headers()) do
		for i, v in pairs(values) do
			if headers[header] == nil then
				headers[header] = v
			else
				headers[header] = headers[header] .. ", " .. v
			end
		end
	end

	-- Make request to backend.
	local b, c, h = http.request {
		url = "http://" .. addr .. path,
		headers = headers,
		create = create_sock,
		-- Disable redirects, because DNS does not work here.
		redirect = false
	}

	-- Check whether we received a valid HTTP response.
	if b == nil then
		txn:Warning("Failure in auth-request backend '" .. be .. "': " .. c)
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- 2xx: Allow request.
	if 200 <= c and c < 300 then
		txn:set_var("txn.auth_response_successful", true)
		txn:set_var("txn.auth_response_code", c)
	-- 401 / 403: Do not allow request.
	elseif c == 401 or c == 403 then
		txn:set_var("txn.auth_response_code", c)
	-- Everything else: Do not allow request and log.
	else
		txn:Warning("Invalid status code in auth-request backend '" .. be .. "': " .. c)
		txn:set_var("txn.auth_response_code", c)
	end
end, 2)


debian/manpage.1.ex
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2018 root <root@unknown>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH Haproxy-auth-request SECTION "January  8 2018"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
haproxy-auth-request \- program to do something
.SH SYNOPSIS
.B haproxy-auth-request
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B haproxy-auth-request
and
.B bar
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBhaproxy-auth-request\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR bar (1),
.BR baz (1).
.br
The programs are documented fully by
.IR "The Rise and Fall of a Fooish Bar" ,
available via the Info system.


auth_request.lua
-- The MIT License (MIT)
--
-- Copyright (c) 2018 Tim Düsterhus
--
-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the "Software"), to deal
-- in the Software without restriction, including without limitation the rights
-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-- copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:
--
-- The above copyright notice and this permission notice shall be included in all
-- copies or substantial portions of the Software.
--
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.

http = require("socket.http")

--- Monkey Patches around bugs in haproxy's Socket class
-- This function calls core.tcp(), fixes a few methods and
-- returns the resulting socket.
-- @return Socket
function create_sock()
	local sock = core.tcp()

	-- https://www.mail-archive.com/haproxy@formilux.org/msg28574.html
	sock.old_receive = sock.receive
	sock.receive = function(socket, pattern, prefix)
		local a, b
		if pattern == nil then pattern = "*l" end
		if prefix == nil then
			a, b = sock:old_receive(pattern)
		else
			a, b = sock:old_receive(pattern, prefix)
		end
		return a, b
	end

	-- https://www.mail-archive.com/haproxy@formilux.org/msg28604.html
	sock.old_settimeout = sock.settimeout
	sock.settimeout = function(socket, timeout)
		socket:old_settimeout(timeout)

		return 1
	end

	-- https://www.mail-archive.com/haproxy@formilux.org/msg28603.html
	sock.old_connect = sock.connect
	sock.connect = function(socket, host, port)
		return socket:old_connect(host..":", port)
	end

	return sock
end

core.register_action("auth_request", { "http-req" }, function(txn, be, path)
	txn:set_var("txn.auth_response_successful", false)

	-- Check whether the given backend exists.
	if core.backends[be] == nil then
		txn:Alert("Unknown auth_request backend '" .. be .. "'")
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- Check whether the given backend has servers that
	-- are not `DOWN`.
	local addr = nil
	for name, server in pairs(core.backends[be].servers) do
		if server:get_stats()['status'] ~= "DOWN" then
			addr = server:get_addr()
		end
	end
	if addr == nil then
		txn:Warning("No servers available for auth_request backend: '" .. be .. "'")
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- Transform table of request headers from haproxy's to
	-- socket.http's format.
	local headers = {}
	for header, values in pairs(txn.http:req_get_headers()) do
		for i, v in pairs(values) do
			if headers[header] == nil then
				headers[header] = v
			else
				headers[header] = headers[header] .. ", " .. v
			end
		end
	end

	-- Make request to backend.
	local b, c, h = http.request {
		url = "http://" .. addr .. path,
		headers = headers,
		create = create_sock,
		-- Disable redirects, because DNS does not work here.
		redirect = false
	}

	-- Check whether we received a valid HTTP response.
	if b == nil then
		txn:Warning("Failure in auth_request backend '" .. be .. "': " .. c)
		txn:set_var("txn.auth_response_code", 500)
		return
	end

	-- 2xx: Allow request.
	if 200 <= c and c < 300 then
		txn:set_var("txn.auth_response_successful", true)
		txn:set_var("txn.auth_response_code", c)
	-- 401 / 403: Do not allow request.
	elseif c == 401 or c == 403 then
		txn:set_var("txn.auth_response_code", c)
	-- Everything else: Do not allow request and log.
	else
		txn:Warning("Invalid status code in auth_request backend '" .. be .. "': " .. c)
		txn:set_var("txn.auth_response_code", c)
	end
end, 2)
